### ISSUE TITLE
Scalability Issue

### ISSUE THREAD

**Issue by ericquan99** [RussellTsuchida/snefy#2]
Hi team,

I have tried your snefy model with high-dimension multivariate gaussian data (d=5/10), with mean = 0 and PSD covariance, using SquaredNN("Rd", "gauss", "snake", "ident") with three layers of feature_net. However, the resulting NLL seems a bit too high (~15).
I was wondering if SNEFY can be applied to high dimensional data.

**Comment by RussellTsuchida** [comment]
Hi again! 

I would think it should be ok for `d=5` or `d=5`. Which data are you using? Have you tried any other methods (e.g. mixture model or normalising flow), and if so, what kind of NLL are they getting?

**Comment by RussellTsuchida** [comment]
If you can share code here, I'm happy to take a look.

**Comment by ericquan99** [comment]
Hi Russell,
Thanks for the reply!
For generating data, the following code is used, which generates 5000 multivariate gaussian (mean=0, PSD covariance) data with d=5:
```
seed = 42
np.random.seed(seed)
torch.manual_seed(seed)

# Generate complete dataset
num_samples = 5000

A = np.random.rand(5, 5)


def generate_well_conditioned_cov_matrix(dim, min_eigenvalue=1, max_eigenvalue=10):
    # Generate random eigenvalues within the specified range
    eigenvalues = np.random.uniform(min_eigenvalue, max_eigenvalue, dim)
    # Generate a random orthogonal matrix
    Q, _ = np.linalg.qr(np.random.randn(dim, dim))
    # Construct the covariance matrix
    cov_matrix = Q @ np.diag(eigenvalues) @ Q.T
    return cov_matrix


dim = 5  # dimension of the covariance matrix
cov_matrix = generate_well_conditioned_cov_matrix(dim)
print("Well-Conditioned Covariance Matrix:")
print(cov_matrix)


# cov_matrix = np.dot(A, A.T)
mean = np.zeros(5)

data = np.random.multivariate_normal(mean, cov_matrix, num_samples)

```

And for SNEFY model, the following code is used, where we added a 3-layer neural network and slightly modified the original Density class to accept the feature net parameter as in Conditional Density class: 
```
n = 8
d = 5
m = n

squared_nn = SquaredNN(
    "Rd", "gauss", "snake", "ident", d=d, n=n, m=n, num_mix_components=1
).to(device)
feat_net = Mlp([5, 128, 64, 8])
snf = Density(squared_nn, feat_net).to(device)
```

We haven't tried other models but will be working on it, and it would be great if you could help check if any of the setting above has caused any problem in calculating the NLL. We would greatly appreciate any comment :)

**Comment by RussellTsuchida** [comment]
Hi @ericquan99 

I copied your code and ran it using the original SNEFY code. I didn't use your slightly modified Density class, which accepts a feature net parameter. (I am actually not even sure what this means, or what modifications have been made to make use of a feature extractor. This is unconditional density estimation, so not sure what a feature extractor would be required for).

Note that in your example, the eigenvalues of the covariance matrix range from between 1 and 10. This means that the variance of the density you are modelling is quite large. This is certainly easy to be modelled by SNEFY by changing the hyperparameters of SNEFY, however it is easier to accommodate this change of scale in the data by preprocessing the data, which I have done on line 39.

The attached code easily obtains -1.9 as a NLL, and could do much better I'm sure with tweaking. For example, `n=m=8` is almost definitely a sub-optimal choice, and the learning rate can probably be bumped up a fair bit.

[gaussian.zip](https://github.com/user-attachments/files/16465980/gaussian.zip)

I will close this issue for now, but as always, I'm always very happy to help and answer any questions you have! Feel free to keep posting them here or in a separate issue. Thanks again for checking out our work!
